substrate/primitives/trie/src/recorder_ext.rs

```diff
@@ -0,0 +1,163 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Extension for the default recorder.
+
+use crate::{RawStorageProof, StorageProof};
+use alloc::{collections::BTreeSet, vec::Vec};
+use scale_info::TypeInfo;
+use sp_core::{Decode, Encode};
+use trie_db::{RecordedForKey, Recorder, TrieAccess, TrieHash, TrieLayout, TrieRecorder};
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, Debug, TypeInfo)]
+pub enum Error {
+	/// The proof contains duplicate nodes.
+	DuplicateNodes,
+	/// The proof contains unused nodes.
+	UnusedNodes,
+}
+
+pub trait RecorderExt<L: TrieLayout>
+where
+	Self: Sized,
+{
+	fn into_set(self) -> BTreeSet<Vec<u8>>;
+
+	fn into_optimized_raw_storage_proof(self) -> RawStorageProof {
+		// The recorder may record the same trie node multiple times,
+		// and we don't want duplicate nodes in our proofs
+		// => let's deduplicate it by collecting to a BTreeSet first
+		self.into_set().into_iter().collect()
+	}
+}
+
+impl<L: TrieLayout> RecorderExt<L> for Recorder<L> {
+	fn into_set(mut self) -> BTreeSet<Vec<u8>> {
+		self.drain().into_iter().map(|record| record.data).collect::<BTreeSet<_>>()
+	}
+}
```

Comment:
> I don't get why we need this trait?

I added the trait in order to avoid duplicating the logic for `into_optimized_storage_proof()`. Since we needed that both for bridges proofs and key ownership proofs.

> Also the recorder is not recording the same node multiple times.

I'm not very familiar with the recorder, but it seems to duplicate something. For example you can try something like the following in the `recorder_ext` tests:

```
	#[test]
	fn unoptimized_proof() {
		let (mut root, raw_proof) = craft_valid_storage_proof();

		let mut recorder = Recorder::<Layout>::new();
		{
			let proof = StorageProof::new(raw_proof.clone());
			let mut db = proof.into_memory_db();
			let trie = TrieDBMutBuilder::<Layout>::new(&mut db, &mut root)
				.with_recorder(&mut recorder)
				.build();

			trie.get(b"key1").unwrap();
			trie.get(b"key11").unwrap();
		}
		assert_eq!(
			recorder.drain().into_iter().map(|record| record.data).collect::<Vec<_>>().len(),
			2
		);

		let mut recorder = Recorder::<Layout>::new();
		{
			let proof = StorageProof::new(raw_proof);
			let mut db = proof.into_memory_db();
			let trie = TrieDBMutBuilder::<Layout>::new(&mut db, &mut root)
				.with_recorder(&mut recorder)
				.build();

			trie.get(b"key1").unwrap();
			trie.get(b"key11").unwrap();
		}
		assert_eq!(recorder.into_raw_storage_proof().len(), 1);
	}
```

Reply:
Yeah! I was confusing it with the implementation in `Substrate` that de-duplicates! Sorry!